pkg/ccl/streamingccl/streamproducer/event_stream.go

```diff
@@ -0,0 +1,438 @@
+				s.maybeSetError(err)
+			}
+			return err
+		}
+	}
+
+	// Context group responsible for coordinating rangefeed event production with
+	// ValueGenerator implementation that consumes rangefeed events and forwards them to the
+	// destination cluster consumer.
+	s.streamGroup = ctxgroup.WithContext(ctx)
+	s.streamGroup.GoCtx(withErrCapture(func(ctx context.Context) error {
+		return s.streamLoop(ctx, frontier)
+	}))
+
+	// TODO(yevgeniy): Add go routine to monitor stream job liveness.
+	// TODO(yevgeniy): Add validation that partition spans are a subset of stream spans.
+}
+
+// Next implements tree.ValueGenerator interface.
+func (s *eventStream) Next(ctx context.Context) (bool, error) {
+	select {
+	case <-ctx.Done():
+		return false, ctx.Err()
+	case err := <-s.errCh:
+		return false, err
+	case s.data = <-s.streamCh:
+		return true, nil
+	}
+}
+
+// Values implements tree.ValueGenerator interface.
+func (s *eventStream) Values() (tree.Datums, error) {
+	return s.data, nil
+}
+
+// Close implements tree.ValueGenerator interface.
+func (s *eventStream) Close(ctx context.Context) {
+	s.rf.Close()
+	s.acc.Close(ctx)
+
+	if err := s.streamGroup.Wait(); err != nil {
+		// Note: error in close is normal; we expect to be terminated with context canceled.
+		log.Errorf(ctx, "partition stream %d terminated with error %v", s.streamID, err)
+	}
+}
+
+func (s *eventStream) onEvent(ctx context.Context, value *roachpb.RangeFeedValue) {
+	select {
+	case <-ctx.Done():
+	case s.eventsCh <- roachpb.RangeFeedEvent{Val: value}:
+		if log.V(1) {
+			log.Infof(ctx, "onEvent: %s@%s", value.Key, value.Value.Timestamp)
+		}
+	}
+}
+
+func (s *eventStream) onCheckpoint(ctx context.Context, checkpoint *roachpb.RangeFeedCheckpoint) {
+	select {
+	case <-ctx.Done():
+	case s.eventsCh <- roachpb.RangeFeedEvent{Checkpoint: checkpoint}:
+		if log.V(1) {
+			log.Infof(ctx, "onCheckpoint: %s@%s", checkpoint.Span, checkpoint.ResolvedTS)
+		}
+	}
+}
+
+func (s *eventStream) onSpanCompleted(ctx context.Context, sp roachpb.Span) error {
+	checkpoint := roachpb.RangeFeedCheckpoint{
+		Span:       sp,
+		ResolvedTS: s.spec.StartFrom,
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case s.eventsCh <- roachpb.RangeFeedEvent{Checkpoint: &checkpoint}:
+		if log.V(1) {
+			log.Infof(ctx, "onSpanCompleted: %s@%s", checkpoint.Span, checkpoint.ResolvedTS)
+		}
+		return nil
+	}
+}
+
+// makeCheckpoint generates checkpoint based on the frontier.
+func makeCheckpoint(f *span.Frontier) (checkpoint streampb.StreamEvent_StreamCheckpoint) {
+	f.Entries(func(sp roachpb.Span, ts hlc.Timestamp) (done span.OpResult) {
+		checkpoint.Spans = append(checkpoint.Spans, streampb.StreamEvent_SpanCheckpoint{
+			Span:      sp,
+			Timestamp: ts,
+		})
+		return span.ContinueMatch
+	})
+	return
+}
+
+func (s *eventStream) flushEvent(ctx context.Context, event *streampb.StreamEvent) error {
+	data, err := protoutil.Marshal(event)
+	if err != nil {
+		return err
+	}
+
```

Comment:
I'm trying to figure out if I think we should be boxing these bytes in another proto, to add any sort of versioning or anything, leave room for a compression-identifying header or whatever.

I suppose we could just say that the passed spec would ask for a particular stream version / encoding / compression / etc that it expects to decode and then this producer is just obligated to either produce what is requested or refuse with an error, so then don't need to worry about it identifying what we produced in the produced data?

so just leaving this as a naked marshaled proto would be okay?